superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx

```diff
@@ -0,0 +1,145 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React from 'react';
+import { render, screen, act } from 'spec/helpers/testing-library';
+import userEvent from '@testing-library/user-event';
+import { SupersetClient } from '@superset-ui/core';
+import * as Utils from 'src/explore/exploreUtils';
+import DatasourceControl from '.';
+
+const SupersetClientGet = jest.spyOn(SupersetClient, 'get');
+
+let defaultProps: any;
+
+beforeEach(() => {
+  defaultProps = {
+    hovered: false,
+    type: 'DatasourceControl',
+    label: 'Datasource',
+    default: null,
+    description: null,
+    value: '25__table',
+    datasource: {
+      id: 25,
+      database: {
+        name: 'examples',
+      },
+      name: 'channels',
+      type: 'table',
+      columns: [],
+    },
+    validationErrors: [],
+    name: 'datasource',
+    actions: {},
+    isEditable: true,
+    onChange: jest.fn(),
+    onDatasourceSave: jest.fn(),
+  };
+});
+
+test('Should render', () => {
+  render(<DatasourceControl {...defaultProps} />);
+  expect(screen.getByTestId('DatasourceControl')).toBeInTheDocument();
+});
+
+test('Should have elements', () => {
+  render(<DatasourceControl {...defaultProps} />);
+  expect(screen.getByText('channels')).toBeVisible();
+  expect(screen.getByTestId('datasource-menu-trigger')).toBeVisible(); // Error
+});
+
+test('Should open a menu', () => {
+  render(<DatasourceControl {...defaultProps} />);
+
+  expect(screen.queryByText('Edit dataset')).not.toBeInTheDocument();
+  expect(screen.queryByText('Change dataset')).not.toBeInTheDocument();
+  expect(screen.queryByText('View in SQL Lab')).not.toBeInTheDocument();
+
+  userEvent.click(screen.getByTestId('datasource-menu-trigger'));
+
+  expect(screen.queryByText('Edit dataset')).toBeInTheDocument();
+  expect(screen.queryByText('Change dataset')).toBeInTheDocument();
+  expect(screen.queryByText('View in SQL Lab')).toBeInTheDocument();
+});
+
+test('Click on Change dataset option', async () => {
+  SupersetClientGet.mockImplementation(
+    async ({ endpoint }: { endpoint: string }) => {
+      if (endpoint.includes('_info')) {
+        return {
+          json: { permissions: ['can_read', 'can_write'] },
+        } as any;
+      }
+      return { json: { result: [] } } as any;
+    },
+  );
+
+  render(<DatasourceControl {...defaultProps} />, {
+    useRedux: true,
+  });
+  userEvent.click(screen.getByTestId('datasource-menu-trigger'));
+
+  await act(async () => {
+    userEvent.click(screen.getByText('Change dataset'));
+  });
```

Comment:
```suggestion
  userEvent.click(await screen.findByText('Change dataset'));
```
`findBy` methods are a combination of `getBy` queries and `waitFor`.

Reply:
In this case the idea is to wait after the click. The element is already on the screen, the click must be within the `await act`.
this code `userEvent.click(await screen.findByText('Change dataset'));` produces a different result than expected.